### PR TITLE
Added create-new-pm endpoint

### PR DESCRIPTION
**Pull request by roansong**
APIV2 documentation: https://osu.ppy.sh/docs/index.html#create-new-pm

This is the first POST request as far as I can tell; I tried to match the existing code style as much as possible.

### CODE REVIEW

File: ossapi/models.py [hunk 1]
```diff
@@ -845,7 +846,43 @@ def override_types(self):
...
+    channel_id: int
+    content: str
+    is_action: bool
+    message_id: int
+    sender: UserCompact
+    sender_id: int
+    timestamp: str
+
+@dataclass
+class CreateNewPMResponse(Model):
```
Comment (roansong) [inline ossapi/models.py:877]: This definitely needs a better name, but what? `CreateNewPM`?

File: ossapi/ossapiv2.py [hunk 2]
```diff
@@ -195,20 +197,33 @@ def _save_token(self, token):
         with open(self.AUTHORIZATION_TOKEN_FILE, "wb+") as f:
             pickle.dump(token, f)
 
-    def _get(self, type_, url, params={}):
+    def _get(self, type_, url, params=None):
+        params = params or {}
```
Comment (tybug) [inline ossapi/ossapiv2.py:201]: any reason you changed the default arg for `params` from `{}` to `None`?

File: ossapi/ossapiv2.py [hunk 3]
```diff
@@ -195,20 +197,33 @@ def _save_token(self, token):
...
         # TODO this should just be ``if "error" in json``, but for some reason
         # ``self.search_beatmaps`` always returns an error in the response...
         # open an issue on osu-web?
-        if len(json_) == 1 and "error" in json_:
-            raise ValueError(f"api returned an error of `{json_['error']}` for "
-                f"a request to {unquote(r.request.url)}")
+        self._check_json(json_, r.request.url)
+        return self._instantiate_type(type_, json_)
+
+    def _post(self, type_, url: str, data: Optional[dict] = None):
```
Comment (tybug) [inline ossapi/ossapiv2.py:213]: personally I don't like type hinting private methods, especially when it's obvious what the type is, like `url`. Just feels harder to read.

File: ossapi/ossapiv2.py [hunk 4]
```diff
@@ -603,6 +618,18 @@ def changelog_lookup(self,
         return self._get(Build, f"/changelog/{changelog}", params)
 
 
+    # /chat
+    # ---------
+
+    @request
+    def create_new_pm(self, target_id: int, message: str, is_action: bool = False) -> CreateNewPMResponse:
```
Comment (tybug) [inline ossapi/ossapiv2.py:625]: let's rename `target_id` to `user_id`, and `create_new_pm` to `create_pm`

File: ossapi/models.py (hunk 1 above)
Comment (tybug) [reply ossapi/models.py:877]: I think we can pare it down slightly to `CreatePMResponse` but other than that I don't have any bright ideas. Not the worst name in any case

**Review by tybug — CHANGES_REQUESTED** [review]

File: ossapi/ossapiv2.py (hunk 2 above)
Comment (roansong) [reply ossapi/ossapiv2.py:201]: I've been caught out by mutable default arguments before; switching to this format is just habit for me now. We don't mutate the dict anywhere so I'm happy to leave it as it was.

File: ossapi/ossapiv2.py (hunk 4 above)
Comment (roansong) [reply ossapi/ossapiv2.py:625]: I think `target_id` is for if/when this endpoint supports group DMs and/or chat channels. `user_id` is more specific to the PM use case, so I'll update

File: ossapi/ossapiv2.py [hunk 5]
```diff
@@ -791,8 +818,7 @@ def user_beatmaps(self,
         if type_ is UserBeatmapType.MOST_PLAYED:
             return_type = List[BeatmapPlaycount]
 
-        return self._get(return_type, f"/users/{user_id}/beatmapsets/"
-            f"{type_.value}", params)
+        return self._get(return_type, f"/users/{user_id}/beatmapsets/{type_.value}", params)
```
Comment (tybug) [inline ossapi/ossapiv2.py:822]: for the record, this was intentional to avoid going over 80 characters

File: tests/test_endpoints.py [hunk 6]
```diff
@@ -112,3 +112,11 @@ def test_deserialize(self):
...
     def test_deserialize(self):
         api.beatmapset_discussion_votes().votes[0].score
+
+
+class TestCreateNewPM(TestCase):
+    def test_deserialize(self):
+        # Target ID of 2070907 is Tillerino
+        scoped_api = OssapiV2(client_id, client_secret, redirect_uri="http://localhost:9409/",
+                              scopes=["chat.write"], strict=True)
+        scoped_api.create_pm(user_id=2070907, message="Integration test please ignore")
```
Comment (tybug) [inline tests/test_endpoints.py:122]: we don't have a good way to deal with multiple differently scoped clients right now and this breaks in weird ways if we leave it mixed in the tests, I think it's fine to leave out the test for now and we can revisit once we rewrite auth handling

**Comment by tybug** [comment]
rewrote a few things, mainly created a `_reqeust` method to handle all our request types, similar to how `requests` does internally

**Pull request merged**